kansha/card_addons/vote/comp.py

```diff
@@ -0,0 +1,71 @@
+# -*- coding:utf-8 -*-
+#--
+# Copyright (c) 2012-2014 Net-ng.
+# All rights reserved.
+#
+# This software is licensed under the BSD License, as described in
+# the file LICENSE.txt, which you should have received as part of
+# this distribution.
+#--
+
+from nagare import security
+from peak.rules import when
+from nagare.security import common
+
+from kansha.card.comp import Card
+from kansha.board.comp import Board
+from kansha.column.comp import Column
+from kansha.cardextension import CardExtension
+from kansha.board.comp import VOTES_PUBLIC, VOTES_MEMBERS
+
+from .models import DataVote
+
+
+@when(common.Rules.has_permission, "user and perm == 'vote' and isinstance(subject, Board)")
+def has_permission_Board_vote(self, user, perm, board):
+    return ((board.has_member(user) and board.votes_allowed == VOTES_MEMBERS)
+            or ((board.votes_allowed == VOTES_PUBLIC) and user))
+
+
+@when(common.Rules.has_permission, "user and perm == 'vote' and isinstance(subject, Column)")
+def has_permission_Column_vote(self, user, perm, column):
+    return security.has_permissions('vote', column.board)
+
+
+@when(common.Rules.has_permission, "user and perm == 'vote' and isinstance(subject, Card)")
+def has_permission_Card_vote(self, user, perm, card):
+    return security.has_permissions('vote', card.column)
+
+
+class Votes(CardExtension):
+    '''Vote component'''
+
+    LOAD_PRIORITY = 70
+
+    def count_votes(self):
+        '''Returns number of votes for a card'''
+        return DataVote.count_votes(self.card.data)
+
+    def vote(self):
+        '''Add a vote to the current card.
+
+        Remove vote if user has already voted
+        '''
+        security.check_permissions('vote', self.card)
+        if not self.has_voted():
+            user = security.get_user()
+            DataVote(card=self.card.data, user=user.data)
+        else:
+            self.unvote()
+
+    def unvote(self):
+        '''Remove a vote to the current card'''
+        security.check_permissions('vote', self.card)
+        if self.has_voted():
+            user = security.get_user()
+            DataVote.get_vote(self.card.data, user.data).delete()
+
```

Comment:
The logic would be simpler with a `toggle` instead of `vote/unvote/has_voted`. See Kansha v2 prototype.